src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java

```diff
@@ -175,6 +175,35 @@ public boolean isReady(@Nonnull JCloudsSlave slave) {
         public abstract boolean isReady(@Nonnull JCloudsSlave slave);
     }
 
+    public enum BootSource {
```

Comment:
I am wondering if there is a way to turn the `BootSource` into an interface with 2 implementations that would replace the `imageId` field in `SlaveOptions` with `bootSource`. The `SlaveOptions` class would not need to treat the image `id`/`kind` any differently than any other. This being polymorphic, we can avoid those `switch`es on numerous places with a method those subclasses would implement differently effectively hiding the differences of the boot sources for all the client code.